Nkpatel/maxpool yolo v4 support

nilaykp opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/7746

### Problem description
Add support for maxpool for highter filter/channel sizes. 

### What's changed
Support is added for filter lengths > 3 and channels > 256 for supporting YoLoV4

### Checklist
- [x] Post commit CI passes [link](https://github.com/tenstorrent/tt-metal/actions/runs/10791755830)
- [x] Device perf regressions and output report [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/10791757920) 
- [] Model perf regressions and output report [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/10791761965/job/29930450448)
- [x] T3000 frequent tests [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/10791766097)
- [] T3000 demo tests [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/10791782441)

nilaykp committed:
Max pool workign for > 256 chennels.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

mywoodstock commented on tests/ttnn/unit_tests/operations/test_max_pool2d.py:
```diff
@@ -289,3 +289,149 @@ def test_pool_core_nondivis(
+        batch_size=in_n,
+        input_height=in_h,
+        input_width=in_w,
+        reader_patterns_cache=reader_patterns_cache,
```
Use the new version `ttnn.max_pool2d_new`

nkpatel-tt replied:
Done.

nilaykp committed:
Disable debug prints.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Make bigger kernel sizes(upto 16x16) work.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Make all test cases for YoloV4 run
  - limit input cb sizes to requirement.
  - Reduce number of operations for reductions by maxing out rows
    to perform partial reduction operations.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Add support for new max pool API.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Add new kernel files for new max pool API.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Make tests for grayscale work.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Working with 16 rows.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Add support for 16 rows.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Update cpp code to chose between small or large kernel based on rows to be processed.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Add double buffer support back.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

mywoodstock commented on ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:
```diff
@@ -137,11 +137,14 @@ MaxPoolNew::MultiCore::cached_program_t max_pool_2d_multi_core_sharded_with_halo
 
+    auto in_cb_sz = (input_shape[3] * kernel_size_hw_padded) > (conv_op_utils::constants::TILE_HW * 8)
+                        ? (conv_op_utils::constants::TILE_HW * 8)
+                        : input_shape[3] * kernel_size_hw_padded;
```
what is `8` here?

nkpatel-tt replied:
Maximum number of tiles which can be reduced at a time. Updated with proper name and comment for better understaning.

mywoodstock commented on ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:
```diff
@@ -168,6 +171,21 @@ MaxPoolNew::MultiCore::cached_program_t max_pool_2d_multi_core_sharded_with_halo
     auto in_tiled_cb = tt::tt_metal::CreateCircularBuffer(program, all_cores, in_tiled_cb_config);
     log_debug(tt::LogOp, "CB {} :: PS = {}, NP = {}", in_tiled_cb_id, in_tiled_cb_pagesize, in_tiled_cb_npages);
 
+    uint32_t max_pool_partials_cb_id = tt::CB::c_intermed1;  // max_pool partials
```
can this CB creation made conditional for large window sizes only? that will keep the current mem usage same for existing models.

nkpatel-tt replied:
Done.

nilaykp committed:
Upadate constants with proper names for better understanding.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

mywoodstock commented:
Also please rebase and combine all commits into one. And make sure pipelines pass.

nilaykp committed:
Rebase on main.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Merge branch 'main' into nkpatel/maxpool_YoloV4_support
